AddonConfiguration well documented in kyma-project.io (#4758)
kyma-project/kyma: docs/helm-broker/02-02-deep-dive.md

---
title: Architecture deep dive
type: Architecture
---

The diagram and steps describe the Helm Broker workflow in details, including the logic of its inner components, namely the Controller and Broker.

![Architecture deep dive](./assets/hb-deep-dive.svg)

1. The Controller watches for ClusterAddonsConfiguration (CAC) and AddonsConfiguration (AC) custom resources (CRs).
2. The user creates, updates, or deletes CAC or AC custom resources.
3. The Controller fetches and parses the data of all addon repositories defined under the **spec.repositories** field in a given CR. During this step, the Controller does the following:

>   - Analyze fetched addons against errors.

Reviewer: ```suggestion
  - Analyzes fetched addons against errors.
```
Analyzes against errors? Don't you mean that it "checks the fetched addons for errors"?

>   - Check for ID duplications under the **repositories** field.

Reviewer: ```suggestion
  - Checks for ID duplication under the **repositories** field.
```

Author: The plural form is intentional here.

>   - Check for ID conflicts with already registered addons.

Reviewer: ```suggestion
  - Checks for ID conflicts with already registered addons.
```

> 4. The Controller persists fetched addons in the storage.

Reviewer: ```suggestion
4. The Controller saves the fetched addons to the storage.
```

> 5. When the first CR appears, the Controller creates ClusterServiceBroker or ServiceBroker, depending on the type of the CR. The ClusterServiceBroker/ServiceBroker provides information about Broker's proper endpoint to the Service Catalog. This endpoint returns the list of available services. There is always only one ClusterServiceBroker per cluster and one ServiceBroker per Namespace, no matter the number of existing CRs. Whenever the list of offered services changes, the Controller triggers the Service Catalog to fetch services from the ClusterServiceBroker/ServiceBroker.

Author: Needed to edit this suggestion a bit. ClusterServiceBroker or ServiceBroker =/= Broker.

6. The Broker component fetches addons from the storage and exposes them to the Service Catalog.
7. The Service Catalog calls the catalog endpoint of the ClusterServiceBroker/ServiceBroker and creates the Service Classes.

## Update CRs

There are two cases in which you might want to update your CR:

> - Re-fetching addons from a remote server

Author: This is not a full sentence, thus I wouldn't finish it with a dot.

> - Re-fetching addons from a remote server

Reviewer: MMS doesn't say it has to be a sentence to use a period 

![image](https://user-images.githubusercontent.com/29116492/61536553-2dce6380-aa35-11e9-9ce4-bc7db3726822.png)

> - Re-fetching addons from a remote server

Author: [Our guidelines](https://github.com/kyma-project/community/blob/master/guidelines/content-guidelines/formatting.md#ordered-and-unordered-lists) say so, see the second bullet point:
"Punctuate bullet points consistently. If they are sentences, use periods. If they are sentence fragments, do not use periods."

> - Re-fetching addons from a remote server

Reviewer: ok

> - Changing repositories URLs

Author: This is not a full sentence, thus I wouldn't finish it with a dot.

### Re-fetching addons

If you provided changes to your addon on a remote server but the URL did not change, you must re-fetch your changes manually. In such a case, increment the **reprocessRequest** field to explicitly request the reprocessing of already registered and processed CR.

### Changing repositories URLs

If you made any change in your addon's URLs, the update process is triggered automatically and the Controller performs its logic.

## Delete CRs

This is the logic the Controller executes after you delete a given CR:

1. If a given CR is in the **Ready** state, the Controller removes it from the storage.
2. After addons are removed from the storage, the Controller increments the **reprocessRequest** field of all failed CRs that had conflicts with the deleted CR in order to reprocess them.

> 3. The Controller deletes a finalizer from the given CR.

Author: Good point, thanks! Link added.
